poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+ * these shares up for a given number of servers to act as guardians, verifying
+ * promises, and providing proof that guardians have indeed taken out approved
+ * of backing up the promse.
+ *
+ * Terms:
+ *   promiser = the server making the promise. The own who owns this object.
+ *   promised = recipients of the promise (aka the clients of the promiser)
+ *   guardian = another server who receives a share of the promise and can help
+ *              reconstruct it.
+ *
+ * Development note: The guardians, secrets, and signatures arrays should
+ *                   remain synchronized. In other words, the guardians[i],
+ *                   secrets[i], and signatures[i] should all refer to the same
+ *                   server.
+ */
+type Promise struct {
+
+	// The id of the promise. In the format:
+	//   PromiserPublicKey.String() + TimeOfCreation + RandomNumber
+	id string
+
+	// The cryptographic group to use for the private shares.
+	shareSuite abstract.Suite
+	
+	// The minimum number of shares needed to reconstruct the secret.
+	t int
+	
+	// The minimum number of shares needed before the policy can become
+	// active. t <= r
+	r int
+	
+	// The total number of shares to send.
+	n int
+	
+	// The public key of the promiser.
+	pubKey abstract.Point
+	
+	// The public polynomial that is used to verify that a secret share
+	// given did indeed come from the appropriate private key.
+	pubPoly *poly.PubPoly
+	
+	// Primarily for the promised, the number of shares that have been 
+	// revealed so far.
+	numShares int
+	
+	// Primarily for the promised, this contains the shares the promised
+	// have currently obtained from guardians. This is what will be used to
+	// reconstruct the secret.
+	priShares * poly.PriShares
+	
+	// The list of servers who act as guardians of the secret. They will
+	// each hold a secret that can be used to decode the promise. The list
+	// is identified by the public key of the serers.
+	guardians   []abstract.Point
+	
+	// The list of secret shares to be sent to the guardians. They are
+	// encrypted with diffie-hellmen shared secrets between the guardian
+	// and the original server.
+	secrets    []abstract.Secret
+	
+	// A list of signatures validating that a guardian has approved of the
+	// secret share it is guarding.
+	signatures []*PromiseSignature
+}
+
+/* Initializes a new promise to guard a secret.
+ *
+ * Arguments
+ *    priKey   = the secret to be promised.
+ *    t        = the minimum number of shares needed to reconstruct the secret.
+ *    r        = the minimum signatures from guardians needed for the promise to
+ *               be valid.
+ *    guardians = a list of the public keys of servers to act as guardians.
+ *
+ * Returns
+ *   The initialized promise
+ *
+ * Note
+ *   Since shares will be multiplied by Diffie-Hellman keys, they need to be the
+ *   same group as the keys.
+ */
+func (p *Promise) Init(keyPair *config.KeyPair, t, r int,
```

Comment:
Does this KeyPair need to include the private key?  Is this Init function only supposed to be called by the promiser or also by nodes that receive and validate a promise?  If it's only for use by the promiser, it's better to name it accordingly to make that clear.